### PR TITLE
Cancel initiated slot correction - [ENG 414]

### PR DESCRIPTION
**Pull request by varunshankar**
**Proposed changes**:
- Modified the pattern_correction as it is decided to not get a confirmation for the user to proceed with the correction.
- Moved the implementation and the flow to docs

**Status (please check what you already did)**:
- [ ] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/main/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

File: rasa/core/policies/flow_policy.py [hunk 1]
```diff
@@ -695,17 +695,31 @@ def _run_step(
...
-                        self._correct_flow_position(
-                            corrected_slots, previous_flow_step, previous_flow, tracker
-                        )
-                    else:
-                        # TODO: we need to figure out how to actually "undo" the
-                        #    changed slots
-                        pass
+                        if (
+                            tracker.latest_action_name
+                            == "utter_corrected_previous_input"
```
Comment (amn41) [inline rasa/core/policies/flow_policy.py:702]: I'm not totally comfortable with the idea of having hardcoded action names in the policy. (at a minimum I think we should define a constant like [here](https://github.com/RasaHQ/rasa/blob/83c9d0c4de05da3b0e01383603cd1002dbfd97bb/rasa/shared/core/constants.py#L25) 

But those other actions are all internal, whereas this is an utter action. Is there not a more robust way to know that we just corrected a slot?

**Review by amn41 — CHANGES_REQUESTED** [review]

File: rasa/core/policies/flow_policy.py [hunk 2]
```diff
@@ -695,17 +695,31 @@ def _run_step(
...
+                        ):
+                            self._correct_flow_position(
+                                corrected_slots,
+                                previous_flow_step,
+                                previous_flow,
+                                tracker,
+                            )
+                        else:
+                            found_first_event = False
```
Comment (tmbo) [inline rasa/core/policies/flow_policy.py:711]: what's this used for?

File: rasa/core/policies/flow_policy.py [hunk 3]
```diff
@@ -695,17 +695,31 @@ def _run_step(
...
+                            for event in reversed(tracker.events):
```
Comment (tmbo) [inline rasa/core/policies/flow_policy.py:712]: What if there are multiple corrected slots?

Can you describe in general the approach you are trying to aim for? Not 100% sure top of my head what the best solution is to identify the previous set value. Is it processing the events in reverse and picking the second to last slotset for each corrected slot? 🤔

File: rasa/core/policies/flow_policy.py (hunk 1 above)
Comment (varunshankar) [reply rasa/core/policies/flow_policy.py]: ya, totally agree with you. I have now used the intent from the latest user message to check if affirm or deny is chosen.
Please let me know if this approach is fine or if there is a better way to do this.

File: rasa/core/policies/flow_policy.py (hunk 2 above)
Comment (varunshankar) [reply rasa/core/policies/flow_policy.py:721]: This is to fetch the second-to-last slot value for a specific slot from the tracker

```
user: I want to transfer 50 USD
bot: ..
user: ahh scratch that, I want to send 100
bot: Do you want to update your information?
user: No, keep the previous information
```

At this point, the `transfer_money_amount_of_money` slot is already set to 100, to revert back to 50, I search for the previous value. Hence, I used a flag variable to fetch the second-to-last value

Comment (tmbo) [reply rasa/core/policies/flow_policy.py:709]: Ah now I get it yes makes sense. but fails if there is more than one slot int `corrected_slots`

File: rasa/core/policies/flow_policy.py (hunk 1 above)
Comment (amn41) [reply rasa/core/policies/flow_policy.py]: unfortunately I'm not sure that will work. We don't rely on "affirm"/"deny" intents with the command generator anymore

File: rasa/core/policies/flow_policy.py (hunk 2 above)
Comment (varunshankar) [reply rasa/core/policies/flow_policy.py:709]: Is there a possibility of having multiple slot corrections from a single-user message? 

The existing implementation was resetting the `corrected_slots` to `None` in the `pattern_correction` flow

Comment (varunshankar) [reply rasa/core/policies/flow_policy.py]: Thanks a lot, Tom, I have now changed the code. Also, added a test here: https://github.com/RasaHQ/financial-demo-flows-llms/pull/28/files

File: rasa/core/policies/flow_policy.py (hunk 1 above)
Comment (varunshankar) [reply rasa/core/policies/flow_policy.py]: This currently works with the command generator, but as we have agreed on changing to default affirmations. Do we hold this task until we implement the changes that we discussed today? 
new task: https://rasahq.atlassian.net/browse/ENG-453

File: rasa/core/policies/default_flows.yml [hunk 4]
```diff
@@ -72,10 +72,6 @@ flows:
       - id: "1"
         action: utter_corrected_previous_input
       - id: "2"
-        set_slots:
```
Comment (amn41) [inline rasa/core/policies/default_flows.yml:75]: as discussed in sprint planning, let's replace this with a simpler `pattern_correction` flow that doesn't do the `confirm_correction` step. And then copy this current version with the confirmation into the docs, to show users how they can override the default flows if they want to add a confirmation step.

**Pull request closed without merging**